Fix problem with item name dropdown selection

coalest opened the pull request:
<!--Read comments, before committing pull request read checklist again

# Checklist:

- I have performed a self-review of my own code,
- I have commented my code, particularly in hard-to-understand areas,
- I have made corresponding changes to the documentation,
- I have added tests that prove my fix is effective or that my feature works,
- New and existing unit tests pass locally with my changes ("bundle exec rake"),
- Title include "WIP" if work is in progress.
- I acknowledge that I will *not* force push my branch once reviews have started.

-->

Resolves #4547 

### Description
This issue occurs when an ajax call to populate the item names in the dropdown options occurs while the dropdown has been opened. Selecting an option doesn't work because that option no longer exists, so nothing happens (so item selected in the field will be the first item from the list of new options).

To fix this I made two changes:
1. Added an event listener so that whenever a item name dropdown option is selected, if the id of that item matches in the list of new options, select that option. (Note: This is enough to fix the bug) 
2. But during investigation I found that when adding a new item, two ajax calls are fired off (to `/inventory.json`) every time you click "Add Another Item". And an ajax call is made to fetch the inventory whenever the storage location changes. I'm assuming it's not needed to fetch the inventory so often. So I changed it to only fetch the inventory item names/quantities on storage location field change, then we store those dropdown options and reuse them whenever a new line item is added.

This will make adding new line items more responsive, and with less jank (from the reduced number of ajax calls) (ie. the item name changing from "XL Diapers" to "XL Diapers (1353)").

### Type of change

* Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?

Tested manually in Chrome on my local machine, using the reproduction steps provided [here](https://github.com/rubyforgood/human-essentials/issues/4547#issuecomment-2242087836).

Testing has been tough. I see a similar issue to the one I'm having where the [`input`event isn't being triggered by cuprite](https://github.com/rubycdp/cuprite/issues/223). In my case the `select2:select` [event](https://select2.org/programmatic-control/events) isn't being triggered for some reason.

dorner commented:
Looks great on the technical side - @cielf did you want to test this out manually?

coalest commented:
Actually, I might I have a better way to resolve this issue. Instead of using an the `select2:select` event listener to detect when an option has been selected and match it to the new option, I think we can might be able change the options in the open dropdown as well. Need to test it out though.

cielf commented:
I would - but now I'll wait.

coalest commented:
So I tried to update the opened dropdown options, but I couldn't find a way to do so using select2's api. I tested out a hacky way (closing and opening the dropdown to refresh), and it kinda works but is a little glitchy if the user is currently scrolling or typing something in dropdown.

In summary, we can fix the bug by matching the options (option 1), or by rerendering the options with some jankiness (option 2).

Here are two videos showing the difference from a user standpoint. I used Chrome and changed the network setting to have 3 seconds of latency.

Video of what option 1 looks like (commit [71698f4](https://github.com/rubyforgood/human-essentials/pull/4797/commits/71698f4f76005e1dd4acd4c86ec4ce58b3bf66db)):
I click "Kids (Size 6)" and it becomes "Kids (Size 6) (1986)".

[Screencast from 2024-11-24 19-35-28.webm](https://github.com/user-attachments/assets/01b3f1cb-c89d-4d3a-a5c5-8b8c756236fd)

Video of what option 2 looks like (commit [bdd3a6a](https://github.com/rubyforgood/human-essentials/pull/4797/commits/bdd3a6aa57d9ef3d83a18ce7dc7d57cfd0a4e819)):
Quantities are added to the dropdown with a bit of jank.
[Screencast from 2024-11-24 19-37-16.webm](https://github.com/user-attachments/assets/13762072-379f-42db-8e35-57df94f3a971)

@dorner @cielf Do y'all have a preference from a functional standpoint?

P.S. Sorry for the force pushing, I had an annoying issue with Rails serving stale JS assets and while debugging it I made some weird commits that I cleaned up.

cielf commented:
I think it's more smooth for the users to see the number of items available as they are picking them, so option # 2, please.

coalest commented:
Updated to use the workaround for option 2. This is now ready for functional and technical review

cielf commented:
LGTM from function pov

dorner commented:
@cielf FYI - the changes in this PR basically completely rewrite the distributions/transfers dropdown logic. Might be worth some extra cycles to regression test it before merging.

cielf commented:
Agreed.  Thanks for pointing that out.

coalest commented:
@dorner If we want to play it safe for production, I can split this into two PRs (one to refresh dropdowns and fix the bug and one with the refactoring).

cielf commented:
I did find one thing.   Here's the sequence:
sign in as org_admin_1@example.com
click on Requests
click on view
click Fulfill request
Pick a storage location
Make one of the items have a bigger quantity than is allowed  (i.e. we are forcing an error)
Save
Yes, it's correct

The quantities for the items disappear.

cielf reviewed (changes requested):
Mostly working like a charm -checked transfers, distributions, and inventory adjustments but found one thing that needs to be addressed on the sequence for distributions made from requests,  if you have an error.  

  Here's the sequence:
sign in as org_admin_1@example.com
click on Requests
click on view
click Fulfill request
Pick a storage location
Make one of the items have a bigger quantity than is allowed  (i.e. we are forcing an error)
Save
Yes, it's correct

The quantities for the items disappear.

coalest commented:
@cielf Good catch. I didn't think about the pages with turbo streams, those aren't triggering the Javascript that is supposed to execute on page load.

The same thing happens when saving a new distribution with errors. (Distributions index => New distribution => Select storage location =>  Save with an error => Quantities disappear)

Does this work currently in production? I feel like this issue should be occurring on the `main` branch as well.

cielf commented:
@coalest Yup -- confirmed  It's happening on staging.    Do you think you can fix it within the bounds of this PR, or do we need to split it up?

coalest commented:
@cielf I just pushed a commit on this branch that should fix the issue.

Ideally, I would like to refactor things a little bit. I don't like the current flow on some pages. Currently (like when we get a request to show the page to edit a distribution), we have access to what the dropdown options should be on the backend if there is a storage location selected (item names + quantities). But instead of providing the HTML with those dropdown options included, we send back default options (with missing quantities) and an ajax request gets triggered via javascript to update labels.

Not a big deal, but I think it could be improved. But I think that could be a future PR.

cielf commented:
@dorner I'm taking this out for a spin,  but could you do another technical pass in parallel?  Thx.

cielf reviewed (commented):
LGTM.   Thanks!

dorner commented:
@coalest if we can separate out the behavior changes from the refactor, I'd much rather do that.

coalest commented on app/javascript/utils/distributions_and_transfers.js:
```diff
@@ -3,69 +3,73 @@ import $ from 'jquery';
-  content += "</option>\n";
-  return content;
+function newOption(item, selectedValue, includeQuantity) {
+  const text = includeQuantity ? item.item_name + ` (${item.quantity})` : item.item_name;
```
Very true, not sure what happened there. Will fix.

coalest commented on spec/system/distribution_system_spec.rb:
```diff
@@ -699,6 +699,34 @@
     end
   end
 
+  context "with dropdown options changed while open", skip: "select2 not working correctly with cuprite" do
```
Really don't understand why I can't get this system spec to work. These steps work in a browser locally. I notice some weird behavior with select2 though. It seems like those events aren't being triggered in cuprite.

coalest replied:
I should probably just delete this?

dorner replied:
@coalest if you can't get the spec working, it's probably best to delete it, yes. Otherwise it'll sit around forever. :(

coalest replied:
Deleted it, as I've already spent a long time trying to get it work. I suspect it's related to that open cuprite issue I linked about, but haven't confirmed.

coalest commented:
Tests aren't passing due to a flaky test. Here is a PR to fix that: #4816

cielf commented:
Huh.   I could have sworn this was working -- but now it isn't,  with the case of fulfilling  a request, picking a storage location,  putting a riduculously high number in a quantity,  saving, confirming, Comes back with an error as expected, but the inventory values do not appear.

cielf reviewed (dismissed):
Huh. I could have sworn this was working earlier -- but now it isn't, with the case of fulfilling a request, picking a storage location, putting a riduculously high number in a quantity, saving, confirming, Comes back with an error as expected, but the inventory values do not appear.

Same with the case of an error on a new distribution.

coalest commented:
@cielf Yea, I took that fix out because I thought we were trying to keep each PR small. so this is now just a PR to fix #4547. Not anything else.

I will add that back in.

coalest commented:
@cielf I put that change back in. So it's ready for review again. Sorry for the confusion.

cielf commented:
(nods) it's still part of  fixing the item dropdown selection, imo.   I can see where it might not have been clear where to draw the line, though.
